returnn/datasets/distrib_files.py

```diff
@@ -151,8 +153,27 @@ def __init__(
             distributed training scenaria
         :param _meta_info_cache: for internal use
         :param _distrib_info: for internal use
+        :param _num_shards: how many data shards there are
+        _param _shard_index: the index of the local shard
         """
-        super().__init__(**kwargs)
+        self.distrib_shard_files = distrib_shard_files
+        if distrib_shard_files:
+            if _distrib_info:
+                # If we're in a child process `_get_rank_and_size()` no longer works,
+                # so we pass the info about the shards via a pickled property.
+                # See also Dataset.__reduce__.
+                shard_index = _distrib_info["_shard_index"]
+                num_shards = _distrib_info["_num_shards"]
+            else:
+                shard_index, num_shards = _get_rank_and_size()
+        else:
+            num_shards = _num_shards or 1
+            shard_index = _shard_index or 0
+        super().__init__(**kwargs, _num_shards=num_shards, _shard_index=shard_index)
+        assert not distrib_shard_files or (_num_shards == 1 and _shard_index == 0), (
```

Comment:
The check is still wrong. You need to check `self._num_shards`. But actually I meant to move the whole code block, so there is nothing above `super().__init__()`.